### PR TITLE
[YUNIKORN-2631] Support canonical labels for queue/applicationId in Admission Controller

### PR DESCRIPTION
**Pull request by chenyulin0719**
### What is this PR for?

When Admission Controller try to patch queue/applicationID to pod, also add canonical representations.

- applicationId (legacy, should be removed in 1.7.0)
- queue  (legacy, should be removed in 1.7.0)
- yunikorn.apache.org/app-id **(New, Canonical Label)**
- yunikorn.apache.org/queue  **(New, Canonical Label)**

In 1.6.0, if only canonical labels were found in pod, add the value back to legacy labels for backward compatibility
- applicationId 
- queue 

More detail could be found in YUNIKORN-1352, [[Design Doc][v4] Define policy for when to use annotations vs labels.pdf  ](https://issues.apache.org/jira/secure/attachment/13067758/%5BDesign%20Doc%5D%5Bv4%5D%20Define%20policy%20for%20when%20to%20use%20annotations%20vs%20labels%20.pdf)
(‘What will be changed’ → ‘Admission Controller Changes)

### What type of PR is it?
* [x] - Improvement

### Todos

- Support canonical representation in scheduler
- Update documents

### What is the Jira issue?
https://issues.apache.org/jira/browse/YUNIKORN-2631

### How should this be tested?
1. make test
2. create below pod and check the pod labels
- pod-with-legacy-labels: AM won't add new labels
- pod-with-no-labels: AM add both legacy/canonical labels
- pod-with-canonical-labels: Besides the existing canonical labels, AM also add the value to legacy labels.
```
apiVersion: v1
kind: Pod
metadata:
  labels:
    app: sleep
    applicationId: "application-sleep-0001"
    queue: "root.sandbox"
  name: pod-with-legacy-labels
spec:
  schedulerName: yunikorn
  restartPolicy: Never
  containers:
    - name: sleep-6000s
      image: "alpine:latest"
      command: ["sleep", "6000"]
      resources:
        requests:
          cpu: "100m"
          memory: "500M"
---
apiVersion: v1
kind: Pod
metadata:
  labels:
    app: sleep
  name: pod-with-no-labels
spec:
  schedulerName: yunikorn
  restartPolicy: Never
  containers:
    - name: sleep-6000s
      image: "alpine:latest"
      command: ["sleep", "6000"]
      resources:
        requests:
          cpu: "100m"
          memory: "500M"
---
apiVersion: v1
kind: Pod
metadata:
  labels:
    app: sleep
    yunikorn.apache.org/app-id: "application-sleep-0003"
    yunikorn.apache.org/queue: "root.sandbox"
  name: pod-with-canonical-labels
spec:
  schedulerName: yunikorn
  restartPolicy: Never
  containers:
    - name: sleep-6000s
      image: "alpine:latest"
      command: ["sleep", "6000"]
      resources:
        requests:
          cpu: "100m"
          memory: "500M"


```

### Screenshots (if appropriate)

- pod-with-legacy-labels
image

- pod-with-no-labels
image

- pod-with-canonical-labels:
image

### Questions:
NA

### CODE REVIEW

File: pkg/common/constants/constants.go [hunk 1]
```diff
@@ -36,8 +36,10 @@ const DomainYuniKornInternal = siCommon.DomainYuniKornInternal
 
 // Application
 const LabelApp = "app"
+const CanonicalLabelApplicationID = DomainYuniKorn + "app-id"
```
Comment (chia7712) [inline pkg/common/constants/constants.go:39]: Could we move `CanonicalLabelApplicationID` to line#42 (after `AnnotationApplicationID`)?

File: pkg/common/constants/constants.go [hunk 2]
```diff
@@ -36,8 +36,10 @@ const DomainYuniKornInternal = siCommon.DomainYuniKornInternal
...
 const LabelApplicationID = "applicationId"
 const AnnotationApplicationID = DomainYuniKorn + "app-id"
+const CanonicalLabelQueueName = DomainYuniKorn + "queue"
```
Comment (chia7712) [inline pkg/common/constants/constants.go:42]: Could it be moved to line#46 (after `AnnotationQueueName`) ?

File: pkg/admission/util.go [hunk 3]
```diff
@@ -37,26 +37,40 @@ func updatePodLabel(pod *v1.Pod, namespace string, generateUniqueAppIds bool, de
...
+	if canonicalAppID == "" && sparkAppID == "" && appID == "" {
 		// if app id not exist, generate one
 		// for each namespace, we group unnamed pods to one single app - if GenerateUniqueAppId is not set
 		// if GenerateUniqueAppId:
 		//		application ID convention: ${NAMESPACE}-${POD_UID}
 		// else
 		// 		application ID convention: ${AUTO_GEN_PREFIX}-${NAMESPACE}-${AUTO_GEN_SUFFIX}
 		generatedID := utils.GenerateApplicationID(namespace, generateUniqueAppIds, string(pod.UID))
+
+		result[constants.CanonicalLabelApplicationID] = generatedID
```
Comment (chia7712) [inline pkg/admission/util.go:51]: Do we need to apply `constants.CanonicalLabelApplicationID` to `GetApplicationIDFromPod`? It seems `GetApplicationIDFromPod` don't take app id from this new label

File: pkg/admission/util.go [hunk 4]
```diff
@@ -37,26 +37,40 @@ func updatePodLabel(pod *v1.Pod, namespace string, generateUniqueAppIds bool, de
...
 	// if existing label exist, it takes priority over everything else
-	if _, ok := existingLabels[constants.LabelQueueName]; !ok {
+	canonicalQueueName := utils.GetPodLabelValue(pod, constants.CanonicalLabelQueueName)
+	queueName := utils.GetPodLabelValue(pod, constants.LabelQueueName)
+	if canonicalQueueName == "" && queueName == "" {
 		// if defaultQueueName is "", skip adding default queue name to the pod labels
 		if defaultQueueName != "" {
 			// for undefined configuration, am_conf will add 'root.default' to retain existing behavior
 			// if a custom name is configured for default queue, it will be used instead of root.default
+			result[constants.CanonicalLabelQueueName] = defaultQueueName
```
Comment (chia7712) [inline pkg/admission/util.go:67]: ditto. should `GetQueueNameFromPod` recognize it?

**Review by chia7712 — COMMENTED** [review]
@chenyulin0719 good to see this patch!

File: pkg/admission/util.go (hunk 3 above)
Comment (chenyulin0719) [reply pkg/admission/util.go:51]: Hi @chia7712, thanks for the review.  
I'm going to change GetApplicationIDFromPod() and GetQueueNameFromPod() in [YUNIKORN-2504 ](https://issues.apache.org/jira/browse/YUNIKORN-2504).  This PR is focus on Admission Controller change. These two function are not called in admission controller.

Comment (chenyulin0719) [reply pkg/admission/util.go:51]: Base on your feedback. I think there have a bug in current Admission Controller and I believe it should be fixed in another PR.
-[ [YUNIKORN-2636] Admission Controller ignores existing Queue/ApplicationID annotations ](https://issues.apache.org/jira/browse/YUNIKORN-2636)

File: pkg/common/constants/constants.go (hunk 1 above)
Comment (chenyulin0719) [reply pkg/common/constants/constants.go]: Done

File: pkg/admission/util.go (hunk 3 above)
Comment (chenyulin0719) [reply pkg/admission/util.go:51]: Merged with the changes in YUNIKORN-2636. 

Based on the same reason in YUNIKORN-2636's discussion.
- https://github.com/apache/yunikorn-k8shim/pull/848#discussion_r1611634390

I'm not going to chang to use GetApplicationIDFromPod()/GetQueueNameFromPod() in Admission Controller.

**Comment by chia7712** [comment]
@chenyulin0719 could you please rebase code to trigger QA again?

**Comment by chenyulin0719** [comment]
The CI fails is the same issue as [[YUNIKORN-2629] Adding a node can result in a deadlock](https://issues.apache.org/jira/browse/YUNIKORN-2629) .

**Review by pbacsko — APPROVED** [review]
+1 LGTM

**Review by chia7712 — APPROVED** [review]
LGTM

**Pull request closed without merging**